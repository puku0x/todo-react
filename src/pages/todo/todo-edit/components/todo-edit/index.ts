export * from './todo-edit.component';
