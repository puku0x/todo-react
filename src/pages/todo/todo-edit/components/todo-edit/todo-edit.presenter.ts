import { useMemo } from 'react';
import { useFormik } from 'formik';
import * as Yup from 'yup';

import { Todo, TodoUpdateDto } from '../../../../../models';

interface FormValues {
  title: string;
  completed: boolean;
}

const toDto = (todo: Todo, values: FormValues) => {
  const value: TodoUpdateDto = {
    id: todo.id,
    title: values.title,
    completed: values.completed,
  };
  return value;
};

export const useTodoEditPresenter = (arg: {
  todo: Todo | null;
  onUpdate?: (todo: TodoUpdateDto) => void;
}) => {
  const { todo, onUpdate } = arg;

  const initialValues = useMemo(() => {
    return {
      title: todo?.title ?? '',
      completed: todo?.completed ?? '',
    } as FormValues;
  }, [todo]);

  const validationSchema = useMemo(() => {
    return Yup.object().shape({
      title: Yup.string().required(),
    });
  }, []);

  const formik = useFormik({
    enableReinitialize: true,
    initialValues,
    validateOnMount: true,
    validationSchema,
    onSubmit: (values) => {
      if (todo) {
        onUpdate?.(toDto(todo, values));
      }
    },
  });

  return {
    ...formik,
  } as const;
};
