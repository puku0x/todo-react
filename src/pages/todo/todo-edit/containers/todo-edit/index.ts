export * from './todo-edit.container';
