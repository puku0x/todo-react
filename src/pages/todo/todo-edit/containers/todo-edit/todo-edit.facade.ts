import { unwrapResult } from '@reduxjs/toolkit';
import { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { TodoUpdateDto } from '../../../../../models';
import {
  fetchTodo,
  updateTodo,
  isFetchingSelector,
  todoSelector,
} from '../../../../../store/todo';

export const useTodoEditFacade = (arg: { id: string }) => {
  const { id } = arg;
  const history = useHistory();
  const dispatch = useDispatch();
  const isFetching = useSelector(isFetchingSelector);
  const todo = useSelector(todoSelector);

  const fetch = useCallback(
    (arg: { id: string }) => {
      return dispatch(fetchTodo(arg)).then(unwrapResult);
    },
    [dispatch]
  );

  const update = useCallback(
    (id: string, dto: TodoUpdateDto) => {
      return dispatch(updateTodo({ id, todo: dto }))
        .then(unwrapResult)
        .then((payload) => {
          const { todo } = payload;
          history.push(`/todos/${todo.id}`);
        });
    },
    [history, dispatch]
  );

  useEffect(() => {
    fetch({ id });
  }, [id, fetch]);

  return {
    isFetching,
    todo,
    fetch,
    update,
  } as const;
};
