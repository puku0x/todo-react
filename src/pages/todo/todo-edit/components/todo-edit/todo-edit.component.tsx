import { FunctionComponent, memo } from 'react';
import { NavLink } from 'react-router-dom';

import { Todo, TodoUpdateDto } from '../../../../../models';
import { useTodoEditPresenter } from './todo-edit.presenter';

type Props = {
  isFetching?: boolean;
  todo: Todo | null;
  onUpdate?: (todo: TodoUpdateDto) => void;
};

export const TodoEdit: FunctionComponent<Props> = memo((props) => {
  const { isFetching, todo, onUpdate } = props;

  const {
    isValid,
    handleBlur,
    handleChange,
    handleSubmit,
  } = useTodoEditPresenter({ todo, onUpdate });

  return (
    <>
      <NavLink to={`/todos/${todo?.id}`}>Back to detail</NavLink>
      <h2>todo-edit</h2>
      {todo ? (
        <>
          <form onSubmit={handleSubmit}>
            <p>
              <button
                type="submit"
                disabled={isFetching || !isValid}
                onBlur={handleBlur}
                onChange={handleChange}
              >
                Save
              </button>
            </p>
            <table>
              <tbody>
                <tr>
                  <td>title</td>
                  <td>
                    <input
                      type="text"
                      name="title"
                      defaultValue={todo.title}
                      onBlur={handleBlur}
                      onChange={handleChange}
                    />
                  </td>
                </tr>
                <tr>
                  <td>completed</td>
                  <td>
                    <input
                      type="checkbox"
                      name="completed"
                      defaultChecked={todo.completed}
                      onBlur={handleBlur}
                      onChange={handleChange}
                    />
                  </td>
                </tr>
              </tbody>
            </table>
          </form>
        </>
      ) : (
        <>loading...</>
      )}
    </>
  );
});
