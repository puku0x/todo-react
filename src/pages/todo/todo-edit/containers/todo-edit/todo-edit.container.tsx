import { memo } from 'react';

import { TodoEdit } from '../../components';
import { useTodoEditFacade } from './todo-edit.facade';

interface Props {
  id: string;
}

export const TodoEditContainer = memo((props: Props) => {
  const { id } = props;
  const { isFetching, todo, update } = useTodoEditFacade({ id });

  return <TodoEdit isFetching={isFetching} todo={todo} onUpdate={update} />;
});
