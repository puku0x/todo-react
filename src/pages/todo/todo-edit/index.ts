import { lazy } from 'react';

export const TodoEditPage = lazy(() =>
  import('./todo-edit.page').then((m) => ({ default: m.TodoEditPage }))
);
