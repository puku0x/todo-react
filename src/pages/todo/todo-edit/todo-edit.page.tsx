import { FunctionComponent, memo } from 'react';
import { useParams } from 'react-router-dom';

import { TodoEditContainer } from './containers';

interface RouterParams {
  id: string;
}

export const TodoEditPage: FunctionComponent = memo(() => {
  const { id } = useParams<RouterParams>();

  return <TodoEditContainer id={id} />;
});
