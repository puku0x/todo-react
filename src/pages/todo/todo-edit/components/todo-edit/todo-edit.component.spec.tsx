import { act, render } from '@testing-library/react';
import { Fragment } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { Todo } from '../../../../../models';
import { TodoEdit } from './todo-edit.component';

describe('TodoEdit', () => {
  it('render', async () => {
    const todo: Todo = {
      id: '1',
      title: 'title',
      completed: false,
      createdAt: 123456789,
      updatedAt: 123456789,
    };
    let result = render(<Fragment />);
    await act(async () => {
      result = render(<TodoEdit todo={todo} />, {
        wrapper: MemoryRouter,
      });
    });
    const { asFragment } = result;

    expect(asFragment()).toMatchSnapshot();
  });

  it('render with isFetching', async () => {
    let result = render(<Fragment />);
    await act(async () => {
      result = render(<TodoEdit isFetching todo={null} />, {
        wrapper: MemoryRouter,
      });
    });
    const { asFragment } = result;

    expect(asFragment()).toMatchSnapshot();
  });
});
