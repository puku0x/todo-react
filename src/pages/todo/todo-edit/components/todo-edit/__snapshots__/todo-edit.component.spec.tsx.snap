// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoEdit render 1`] = `
<DocumentFragment>
  <a
    href="/todos/1"
  >
    Back to detail
  </a>
  <h2>
    todo-edit
  </h2>
  <form>
    <p>
      <button
        type="submit"
      >
        Save
      </button>
    </p>
    <table>
      <tbody>
        <tr>
          <td>
            title
          </td>
          <td>
            <input
              name="title"
              type="text"
              value="title"
            />
          </td>
        </tr>
        <tr>
          <td>
            completed
          </td>
          <td>
            <input
              name="completed"
              type="checkbox"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </form>
</DocumentFragment>
`;

exports[`TodoEdit render with isFetching 1`] = `
<DocumentFragment>
  <a
    href="/todos/undefined"
  >
    Back to detail
  </a>
  <h2>
    todo-edit
  </h2>
  loading...
</DocumentFragment>
`;
